assessment/libs/stats.php

```diff
@@ -2401,4 +2402,420 @@ function dotplot($a,$label,$dotspace=1,$labelspace=null,$width=300,$height=150)
+	
+	$n=array();  
+	foreach($arr as $a){
+		if (!is_array($a)) { $a = explode(',',$a);};
+		$n[]=count($a);
+		/*if (count(array_unique($n))!=1) {
+			echo "Error: ANOVA requires the same length for all arrays";
+			return false;
+		}*/
+	}
+	if (count($n)<3) {
+		echo "Error: ANOVA requires three or more arrays";
+		return false;
+	}
+	$N=array_sum($n);	
+	$numargs = func_num_args();
+	//$args=func_get_args();
+	$mean=array();
+	$ss=array();
+	for($i=0;$i<$numargs;$i++){
+		$mean[$i]=mean($arr[$i]);
+		$ss[$i]=variance($arr[$i])*(count($arr[$i])-1);
+		//$n[$i]=count($args[$i]);
+	}
+	for ($i=0;$i<$numargs;$i++){
+		$gmean=array_sum($mean)/count($mean); //grand mean 
+	}
+	$total = array_map(function($x, $y) { return $x * $y; },
+                   $mean, $n);
+
+	$gmean=array_sum($total)/$N; //grand mean
+
+	//Sum of the square for Factor A uneequal sample sizes
+	$ssa=array();
+	for ($i=0;$i<$numargs;$i++){
+		$ssa[$i]=$n[$i]*($mean[$i]-$gmean)**2; 
+	}
+	$ssA=array_sum($ssa); //Sum of the square for Factor A uneequal sample sizes
+	$ssE=array_sum($ss); //Sum of the square for Residual (Error)
+	$ssT=$ssA+$ssE;
+	$dfA=count($arr)-1;
+	$dfE=array_sum($n)-$numargs;
+	$dfT=$dfA+$dfE;
+	
+	
+	$msA=$ssA/$dfA; //mean square of Factor A
+	$msE=$ssE/$dfE;    //pooled variance (residual or error)
+	$msT=$msA+$msE;  //total sum of the squares
+	$F_a=$msA/$msE;    //F Ratio
+	
+	$p_a=fcdf($F_a,$dfA,$dfE); //P value
+	
+	return ([$F_a,$p_a]); 
+}
+
+
+//---------------------------------------------ANOVA-Oneway array-------------------------------------------
+// Function: anova1way(arr1,arr2, [arr3,...])
+// Returns ANOVA table as an array with each row corresponding to Factor A, error (residual), and totals. 
+//
+// Parameters:
+// arr1, arr2, ...: Arrays in the form [2,3,4,5,...]; it also accepts unequal sample sizes. 
+//  
+// Returns:
+// ANOVA table as an array in the following format. This array can be used in anova_table() to tabulate data for display.
+// [[SS_A, df_A, MS_A, F_A, P_A],[SS_E,df_E,MS_E],[SS_T,df_T]] 
+// where SS is sum of the squares, df is the degree of freedom, MS is mean square, F is F ratio, and P is P value.
+// And A, E, and T correspond to Factor A, error (residual), and total, respectively. 
+
+function anova1way(... $arr){
+	
+	$n=array();  
+	foreach($arr as $a){
+		if (!is_array($a)) { $a = explode(',',$a);};
+		$n[]=count($a);
+		/*if (count(array_unique($n))!=1) {
+			echo "Error: ANOVA requires the same length for all arrays";
+			return false;
+		}*/
+	}
+	if (count($n)<3) {
+		echo "Error: ANOVA requires three or more arrays";
+		return false;
+	}
+	$N=array_sum($n);	
+	$numargs = func_num_args();
+	//$args=func_get_args();
+	$mean=array();
+	$ss=array();
+	for($i=0;$i<$numargs;$i++){
+		$mean[$i]=mean($arr[$i]);
+		$ss[$i]=variance($arr[$i])*(count($arr[$i])-1);
+		//$n[$i]=count($args[$i]);
+	}
+	for ($i=0;$i<$numargs;$i++){
+		$gmean=array_sum($mean)/count($mean); //grand mean 

```

Comment:
Ditto on above

Reply:
Oops. Yes, that doesn't do anything. Thanks for catching that.